bin/lib/invalidUsernames.js

```diff
@@ -0,0 +1,160 @@
+const Handlebars = require('handlebars')
+const path = require('path')
+const { URL } = require('url')
+const util = require('util')
+
+const { loadConfig } = require('./common')
+const { isValidUsername } = require('../../lib/common/user-utils')
+const blacklistService = require('../../lib/services/blacklist-service')
+const { initConfigDir, initTemplateDirs } = require('../../lib/server-config')
+const { fromServerConfig } = require('../../lib/models/oidc-manager')
+
+const AccountManager = require('../../lib/models/account-manager')
+const EmailService = require('../../lib/services/email-service')
+const LDP = require('../../lib/ldp')
+const SolidHost = require('../../lib/models/solid-host')
+
+const fileExists = util.promisify(fs.exists)
+const fileRename = util.promisify(fs.rename)
+
+module.exports = function (program) {
+  program
+    .command('invalidusernames')
+    .option('--notify', 'Will notify users with usernames that are invalid')
+    .option('--delete', 'Will delete users with usernames that are invalid')
+    .description('Manage usernames that are invalid')
+    .action(async (options) => {
+      const config = await loadConfig(program, options)
+      if (!config.multiuser) {
+        return console.error('You are running a single user server, no need to check for invalid usernames')
+      }
+
+      const invalidUsernames = await getInvalidUsernames(config)
+      const host = SolidHost.from({ port: config.port, serverUri: config.serverUri })
+      const accountManager = getAccountManager(config, host)
+
+      if (options.notify) {
+        return notifyUsers(invalidUsernames, accountManager, config)
+      }
+
+      if (options.delete) {
+        return deleteUsers(invalidUsernames, accountManager, config, host)
+      }
+
+      listUsernames(listUsernames)
+    })
+}
+
+async function createNewIndexFile (username, accountManager, invalidUsernameTemplate, dateOfRemoval, supportEmail, fileOptions) {
+  const userDirectory = accountManager.accountDirFor(username)
+  const currentIndex = path.join(userDirectory, 'index.html')
+  const currentIndexExists = await fileExists(currentIndex)
+  const backupIndex = path.join(userDirectory, 'index.backup.html')
+  const backupIndexExists = await fileExists(backupIndex)
+  if (currentIndexExists && !backupIndexExists) {
+    await fileRename(currentIndex, backupIndex)
+    const newIndexSource = invalidUsernameTemplate({
+      username,
+      dateOfRemoval,
+      supportEmail
+    })
+    fs.writeFileSync(currentIndex, newIndexSource, fileOptions)
+    console.info(`index.html updated for user ${username}`)
+  }
+}
+
+async function deleteUsers (usernames, accountManager, config, host) {
+  const oidcManager = fromServerConfig({
+    ...config,
+    host
+  })
+  const deletingUsers = usernames
+    .map(async username => {
+      try {
+        const user = accountManager.userAccountFrom({ username })
+        await oidcManager.users.deleteUser(user)
```

Comment:
IIRC I've read in the docs, that other auth managers (than oidc) would be supported as well.
So I am wondering here, why it has to be oidc.
Maybe add an comment? Could be that it is a „stupid question” for contributors familiar with the code.

Reply:
No, it's not a stupid question; We support WebID-TLS and WebID-OIDC now, and WebID-TLS don't store user-data like WebID-OIDC does.

I could try to abstract away the OIDC-part of this, to accommodate future auth managers better, but I think that's going to might take a long time (for me, at least - I'm not that into auth managers yet). So I was thinking this can be done when we implement support for other auth managers.

Reply:
Maybe you could leave a comment here, so the next person touching the code knows your thoughts (since I'd not look up the associated PR immediately, but rather `git blame` …).

I don't know the conventions in this project, though.